### PR TITLE
Conditionally allow medial searches

### PR DESCRIPTION
**Pull request by jvillafanez**
Conditionally allow medial searches, or search in the middle of strings.

For small LDAP servers you'd only need to enable the feature. For big installations is recommended to index the attributes with the specific index in the LDAP server before enabling this feature (check on your LDAP provider how you can optimize searches like "*user*")

NOTE: This feature is connection-based. If ownCloud is connected to several LDAPs, you'll have to enable (or disable) this feature for each connection in order to have an homogeneous behaviour.

Related issues:
https://github.com/owncloud/core/issues/26519
https://github.com/owncloud/enterprise/issues/1640
https://github.com/owncloud/core/issues/25216

To be checked if we want to backport this enhancement or not.

@PVince81

### CODE REVIEW

File: templates/settings.php [hunk 1]
```diff
@@ -93,6 +93,7 @@
...
 				<p><label for="ldap_nested_groups"><?php p($l->t('Nested Groups'));?></label><input type="checkbox" id="ldap_nested_groups" name="ldap_nested_groups" value="1" data-default="<?php p($_['ldap_nested_groups_default']); ?>"  title="<?php p($l->t('When switched on, groups that contain groups are supported. (Only works if the group member attribute contains DNs.)'));?>" /></p>
+				<p><label for="ldap_medial_searches"><?php p($l->t('Medial Searches'));?></label><input type="checkbox" id="ldap_medial_searches" name="ldap_medial_searches" value="1" data-default="<?php p($_['use_medial_searches_default']); ?>"  title="<?php p($l->t('By default, the searches for users or groups will take into account only the beginning of the string (if you search "al", "alice" will be found, but not "balnar"). Medial searches allows matching in any part of the string (in the previous example, both names would match). Note that some indexes might need to be added in some LDAP servers in order to give a good performance in big installations'));?>" /></p>
```
Comment (PVince81) [inline templates/settings.php]: Is this something a LDAP admin understands directly or does it need more hints or a link to a more detailed documentation ? (from doc repo)

cc @cdamken

**Comment by PVince81** [comment]
Is there a risk of enabled medial search to interfere with other features and cause potential regressions ? Just asking to be sure 😄 

@owncloud/qa please add to a test plan until we can have automated tests for this

File: templates/settings.php (hunk 1 above)
Comment (jvillafanez) [inline templates/settings.php]: This needs to be documented for sure, specially how to setup the indexes in the most common LDAP providers. We can add a hint to the documentation.

**Comment by jvillafanez** [comment]
I don't think so, unless the function does something that it isn't supposed to do. My expectation is that the search will return different results (instead of "a" it will return "a", "b" and "c"), but how the results are handled is something that remains untouched.

**Comment by PVince81** [comment]
@davitol please help testing this manually and add the step on the test plan / todo for automated tests.

**Comment by PVince81** [comment]
:+1: for the code change, needs testing

**Comment by davitol** [comment]
> 👍 for the code change, needs testing

https://github.com/owncloud/QA/issues/369  

So :+1: from QA POV.

@pmaier1 @felixboehm @PVince81 Maybe this 'plus feature' should be added in Platform repo

Note: https://github.com/owncloud/core/issues/27013

File: lib/Access.php [hunk 2]
```diff
@@ -1369,12 +1369,17 @@ private function prepareSearchTerm($term) {
 		$config = \OC::$server->getConfig();
 
 		$allowEnum = $config->getAppValue('core', 'shareapi_allow_share_dialog_user_enumeration', 'yes');
+		$allowMedialSearches = $config->getSystemValue('user_ldap.medial_search.enable', false);
```
Comment (PVince81) [inline lib/Access.php]: you probably saw that there is only one other key called `ldapUserCleanupInterval`.
I'm not too happy about the naming scheme inconsistenc<.

On the other hand, the naming you used makes more sense for the future. So I think it's ok. Please add this to config.sample.php.

Comment (PVince81) [inline lib/Access.php]: Do you think there will be other options related to medial search ? If not, just call it `user_ldap.enable_medial_search`.

Please also add it to config.sample.php and specify that it is experimental

**Comment by jvillafanez** [comment]
> Please also add it to config.sample.php and specify that it is experimental

Does it make sense to document it it in config.sample.php? The app is out of core now, so it would be better to add the documentation somewhere in this new repo.
It's weird if the core config reference something that it isn't longer in core, plus other apps won't have this "advantage"

**Comment by PVince81** [comment]
Hmm indeed, then add it to the README of the user_ldap app

File: README.md [hunk 3]
```diff
@@ -2,3 +2,17 @@
...
 [![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/owncloud/user_ldap/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/owncloud/user_ldap/?branch=master)
 [![Build Status](https://scrutinizer-ci.com/g/owncloud/user_ldap/badges/build.png?b=master)](https://scrutinizer-ci.com/g/owncloud/user_ldap/build-status/master)
+
+#### Aditional configuration options that can be added to config.php
+
+* `'user_ldap.enable_medial_search' => true`
+
+    By default, when you search for a user your input string will match the beginning of the username. For example, if your LDAP server has "erl" and "peter" as users and you search with "er", only "erl" will be shown.
+
+    Enabling this options allows you to overcome this limitation. In the example above, when this option is active, searching for "er" will find both users.
```
Comment (phil-davis) [inline README.md]: "Enabling this option..."

File: README.md [hunk 4]
```diff
@@ -2,3 +2,17 @@
...
+
+    Before enabling this option take into account the following things:
+
+    * This option affects all LDAP connections. It isn't possible to enable this option for a specific connection.
+    * This option could have a performance impact on big LDAP installations. Check your LDAP provider how to enable indexes for medial searches if they're supported but not active.
+    * The option will work regardless of whether the LDAP server has an index for this. Small LDAP installations could have an acceptable performance with this option active even if the LDAP doesn't has that index active in the LDAP server.
```
Comment (phil-davis) [inline README.md]: "even if the LDAP server doesn't have that index active."
IMHO "in the LDAP server" at the end of the sentence is unneeded words.

Comment (scolebrook) [inline README.md]: "... even if the LDAP doesn't **have** that index active ..."

File: README.md [hunk 5]
```diff
@@ -2,3 +2,17 @@
 [![Code Coverage](https://scrutinizer-ci.com/g/owncloud/user_ldap/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/owncloud/user_ldap/?branch=master)
 [![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/owncloud/user_ldap/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/owncloud/user_ldap/?branch=master)
 [![Build Status](https://scrutinizer-ci.com/g/owncloud/user_ldap/badges/build.png?b=master)](https://scrutinizer-ci.com/g/owncloud/user_ldap/build-status/master)
+
+#### Aditional configuration options that can be added to config.php
```
Comment (davitol) [inline README.md]: typo Aditional instead of Additional

**Comment by jvillafanez** [comment]
I guess there won't be any additional change... :shipit:

**Comment by jvillafanez** [comment]
bump!

**Comment by PVince81** [comment]
:+1:

**Comment by jvillafanez** [comment]
Backport to 9.1?

**Comment by PVince81** [comment]
No, it's a like a new feature

**Pull request merged**